File: assignments/assignment2a/ugrad/app/src/main/java/edu/vanderbilt/webcrawler/platform/AndroidImage.java
package edu.vanderbilt.webcrawler.platform;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.util.Log;

import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;

import edu.vanderbilt.platform.Platform;
import edu.vanderbilt.platform.PlatformImage;
import edu.vanderbilt.webcrawler.App;
import edu.vanderbilt.webcrawler.R;

/**
 * Stores platform-specific meta-data about an Image and also provides
 * methods for common image- and file-related tasks.  This
 * implementation is specific to the Android platform.
 */
public class AndroidImage implements PlatformImage {
	/**
	 * Logging tag.
	 */
	private static final String TAG = "AndroidImage";

	/**
	 * Dimensions representing how large the scaled image should be.
	 */
	private static final int IMAGE_WIDTH = 250;
	private static final int IMAGE_HEIGHT = 250;

	/**
	 * Platform helper methods passed in to constructor.
	 */
	private final Platform mPlatform;

	/**
	 * The Bitmap image.
	 */
	private Bitmap mImage;

	/**
	 * The original source of this image.
	 */
	private URL mUrl;

	/**
	 * Package only constructor only accessed by Platform.
	 */
	AndroidImage(URL url, byte[] imageData, Platform platform) {
		mPlatform = platform;
		setImage(url, imageData);
	}

	/**
	 * Private constructor only accessed internally by this class.
	 */
	private AndroidImage(URL url, Bitmap image, Platform platform) {
		mPlatform = platform;
		mUrl = url;
		mImage = image;
	}

	@Override
	public void setImage(URL url, byte[] imageData) {
		mUrl = url;
		if (imageData == null) {
			mImage = BitmapFactory.decodeResource(
					App.Companion.getInstance().getResources(), R.mipmap.ic_launcher_round);
		} else {
			mImage = decodeSampledBitmapFromByteArray(imageData,
					IMAGE_WIDTH,
					IMAGE_HEIGHT);
		}
	}

	@Override
	public URL getSourceUrl() {
		return mUrl;
	}

	@Override
	public String mapUrlToRelativeFilePath() {
		return mPlatform.mapUrlToRelativeFilePath(mUrl);
	}

	/**
	 * Decodes and scales a bitmap from a byte array.  Adapted from
	 * developer.android.com/training/displaying-bitmaps/load-bitmap.html
	 */
	private Bitmap decodeSampledBitmapFromByteArray(
			byte[] imageData, int reqWidth, int reqHeight) {

		// First decode with inJustDecodeBounds=true to check
		// dimensions.
		final BitmapFactory.Options options = new BitmapFactory.Options();
		options.inJustDecodeBounds = true;
		BitmapFactory.decodeByteArray(imageData,
				0,
				imageData.length,
				options);

		// Calculate inSampleSize.
		options.inSampleSize = calculateInSampleSize(options,
				reqWidth,
				reqHeight);

		// Decode bitmap with inSampleSize set.
		options.inJustDecodeBounds = false;
		return BitmapFactory.decodeByteArray(imageData,
				0,
				imageData.length,
				options);
	}

	/**
	 * Calculates the Bitmap's sampling rate to fit the given
	 * dimensions. Adapted from
	 * developer.android.com/training/displaying-bitmaps/load-bitmap.html
	 */
	private int calculateInSampleSize(
			BitmapFactory.Options options, int reqWidth, int reqHeight) {

		// Raw height and width of image.
		final int height = options.outHeight;
		final int width = options.outWidth;
		int inSampleSize = 1;

		if (height > reqHeight || width > reqWidth) {

			final int halfHeight = height / 2;
			final int halfWidth = width / 2;

			// Calculate the largest inSampleSize value that is a
			// power of 2 and keeps both height and width larger than
			// the requested height and width.
			while ((halfHeight / inSampleSize) > reqHeight
					&& (halfWidth / inSampleSize) > reqWidth) {
				inSampleSize *= 2;
			}
		}

		return inSampleSize;
	}

	/**
	 * Write the @a image to the @a outputStream.
	 */
	@Override
	public void writeImageFile(FileOutputStream outputStream)
			throws IOException {
		if (mImage == null) {
			Log.d(TAG, "NULL image!");
		} else {
			mImage.compress(Bitmap.CompressFormat.PNG, 90, outputStream);
			outputStream.flush();
		}
	}

	/**
	 * Apply a grayscale filter to the @a imageEntity and return it.
	 */
	public PlatformImage applyFilter() {
		// Bail out if something is wrong with the image.
		if (mImage == null) {
			return null;
		}

		Bitmap originalImage = mImage;
		Bitmap grayScaleImage =
				originalImage.copy(originalImage.getConfig(), true);

		boolean hasTransparent = grayScaleImage.hasAlpha();
		int width = grayScaleImage.getWidth();
		int height = grayScaleImage.getHeight();

		// A common pixel-by-pixel grayscale conversion algorithm
		// using values obtained from en.wikipedia.org/wiki/Grayscale.
		for (int i = 0; i < height; ++i) {
			// Break out if we've been interrupted.
			if (Thread.interrupted())
				return null;

			for (int j = 0; j < width; ++j) {

				// Check if the pixel is transparent in the original
				// by checking if the alpha is 0
				if (hasTransparent
						&& ((grayScaleImage.getPixel(j, i) & 0xff000000) >> 24) == 0) {
					continue;
				}

				// Convert the pixel to grayscale.
				int pixel = grayScaleImage.getPixel(j, i);
				int grayScale =
						(int) (Color.red(pixel) * .299
								+ Color.green(pixel) * .587
								+ Color.blue(pixel) * .114);
				grayScaleImage.setPixel(j, i, Color.rgb(grayScale,
						grayScale,
						grayScale));
			}
		}

		return new AndroidImage(getSourceUrl(), grayScaleImage, mPlatform);
	}
}

File: assignments/assignment4-final/app/src/main/java/edu/vanderbilt/crawler/platform/AndroidImage.kt
package edu.vanderbilt.crawler.platform

import android.graphics.Bitmap
import android.graphics.Color
import edu.vanderbilt.crawler.R
import edu.vanderbilt.crawler.app.App
import edu.vanderbilt.crawler.utils.BitmapUtils
import edu.vanderbilt.crawler.utils.KtLogger
import edu.vanderbilt.imagecrawler.crawlers.ImageCrawler
import edu.vanderbilt.imagecrawler.platform.Cache
import edu.vanderbilt.imagecrawler.platform.PlatformImage
import edu.vanderbilt.imagecrawler.transforms.Transform
import org.jetbrains.anko.dimen
import java.io.InputStream
import java.io.OutputStream

/**
 * Stores platform-specific meta-data about an Image and also provides
 * methods for common image- and file-related tasks.  This
 * implementation is specific to the Android platform.
 */
class AndroidImage : PlatformImage, KtLogger {
    companion object {
        /** Dimensions representing how large the scaled image should be. */
        private val IMAGE_WIDTH = App.instance.dimen(R.dimen.url_grid_image_size)
        private val IMAGE_HEIGHT = IMAGE_WIDTH

    }

    lateinit var item: Cache.Item

    var image: Bitmap? = null
    var size = 0

    constructor(inputStream: InputStream, item: Cache.Item) {
        setImage(inputStream, item)
    }

    constructor(bitmap: Bitmap) {
        image = bitmap
        size = bitmap.byteCount
    }

    override fun setImage(inputStream: InputStream, item: Cache.Item) {
        this.item = item
        size = inputStream.available()
        image = BitmapUtils.decodeSampledBitmapFromStream(
                inputStream, IMAGE_WIDTH, IMAGE_HEIGHT)
        size = image?.byteCount ?: 0
    }

    /**
     * Returns size of image.
     */
    override fun size(): Int = size

    /**
     * Write the image to the [outputStream].
     */
    override fun writeImage(outputStream: OutputStream) {
        image?.compress(Bitmap.CompressFormat.PNG, 90, outputStream)
    }

    /**
     * Apply a grayscale filter to the @a imageEntity and return it.
     */
    override fun applyTransform(type: Transform.Type,
                                newItem: Cache.Item): PlatformImage? {

        ImageCrawler.throwExceptionIfCancelled();

        return when (type) {
            Transform.Type.GRAY_SCALE_TRANSFORM -> grayScale(newItem)
            Transform.Type.NULL_TRANSFORM -> this
        }
    }

    private fun grayScale(newItem: Cache.Item): PlatformImage? {
        // Bail out if something is wrong with the image.
        if (image == null) {
            return null
        }

        val originalImage = image
        val grayScaleImage = originalImage!!.copy(originalImage.config, true)

        val hasTransparent = grayScaleImage.hasAlpha()
        val width = grayScaleImage.width
        val height = grayScaleImage.height
        val total = width * height
        var bytes = 0

        // A common pixel-by-pixel grayscale conversion algorithm
        // using values obtained from en.wikipedia.org/wiki/Grayscale.
        for (i in 0 until height) {
            for (j in 0 until width) {
                ImageCrawler.throwExceptionIfCancelled();

                bytes++

                // Check if the pixel is transparent in the original
                // by checking if the alpha is 0
                if (hasTransparent && grayScaleImage.getPixel(j, i) and 0xff000000.toInt() shr 24 == 0) {
                    continue
                }

                // Convert the pixel to grayscale.
                val pixel = grayScaleImage.getPixel(j, i)
                val grayScale = (Color.red(pixel) * .299
                        + Color.green(pixel) * .587
                        + Color.blue(pixel) * .114).toInt()
                grayScaleImage.setPixel(j, i, Color.rgb(grayScale,
                        grayScale,
                        grayScale))
            }

            newItem.progress(Cache.Operation.TRANSFORM, bytes.toFloat() / total, bytes)
        }

        return AndroidImage(grayScaleImage)
    }
}
